clang wrapper behaves differently than clang on OSX

This is on MacOS high sierra

balay@ipro^~/spack(xsdk-0.3.0) $ sw_vers 
ProductName:	Mac OS X
ProductVersion:	10.13
BuildVersion:	17A405


I see spack is using the wrapper lib/spack/env/clang/clang instead of /usr/bin/clang

Perhaps there is a valid reason for this. But I have 2 issues with it:

1. It behaves differently than /usr/bin/clang. Why?

```
     17    checking for gcc... /Users/balay/spack/lib/spack/env/clang/clang
     18    checking whether the C compiler works... yes
     19    checking for C compiler default output file name... a.out
     20    checking for suffix of executables...
  >> 21    checking whether we are cross compiling... configure: error: in `/private/var/folders/b2/4wzkzbgs3d91kqp1dw47wq940000gn/T/balay/spack-stage/spack-stage-4h3G7M/pkg-config-0.29.2':
  >> 22    configure: error: cannot run C compiled programs.

```

config.log for this build has:
```
conftest.c:11:10: fatal error: 'stdio.h' file not found
#include <stdio.h>
         ^~~~~~~~~
1 error generated.
```

However /usr/bin/clang works.

```
balay@ipro^~/spack(xsdk-0.3.0) $ cat conftest.c 
#include <stdio.h>
int main(void)
{return 0;}
balay@ipro^~/spack(xsdk-0.3.0) $ clang conftest.c 
balay@ipro^~/spack(xsdk-0.3.0) $ 
```

2. I cannot run this wrapper outside spack to reproduce issues [by manual build of packages or trying to compile test code]

```
balay@ipro^~/spack(xsdk-0.3.0) $ /Users/balay/spack/lib/spack/env/clang/clang conftest.c 
Spack compiler must be run from Spack! Input 'SPACK_PREFIX' is missing.
balay@ipro^~/spack(xsdk-0.3.0) $ export SPACK_PREFIX=$PWD
balay@ipro^~/spack(xsdk-0.3.0) $ /Users/balay/spack/lib/spack/env/clang/clang conftest.c 
Spack compiler must be run from Spack! Input 'SPACK_ENV_PATH' is missing.
balay@ipro^~/spack(xsdk-0.3.0) $ export SPACK_ENV_PATH=$PWD
balay@ipro^~/spack(xsdk-0.3.0) $ /Users/balay/spack/lib/spack/env/clang/clang conftest.c 
Spack compiler must be run from Spack! Input 'SPACK_DEBUG_LOG_DIR' is missing.
balay@ipro^~/spack(xsdk-0.3.0) $ export SPACK_DEBUG_LOG_DIR=$PWD
balay@ipro^~/spack(xsdk-0.3.0) $ /Users/balay/spack/lib/spack/env/clang/clang conftest.c 
Spack compiler must be run from Spack! Input 'SPACK_COMPILER_SPEC' is missing.
balay@ipro^~/spack(xsdk-0.3.0) $ 
```
and so on... How do I invoke this wrapper for manual verification?

 
I see Spack doc says 'xcode-select --install' is required. I guess this will resolve this error.
http://spack.readthedocs.io/en/latest/getting_started.html?highlight=xcode-select

But I'm able to install various packages without it. So Its not clear to me why this wrapper has more requirements than /usr/bin/clang. Previously I've seen this requirement for external compilers - like pgi - or clang installed by brew - but in this case - I'm using clang from XCode.

Note: the following works [when it fails via spack]:
```
balay@ipro^~/spack/var/spack/stage/pkg-config-0.29.2-zggrhmtc22dolyoug5ylkrknom2e3vda/pkg-config-0.29.2 $ CC=/usr/bin/clang ./configure --enable-shared --with-internal-glib

```

Thanks! 'spack env <spec> bash is useful.

Here is the difference in behavior between spack clang and /usr/bin/clang

```
balay@ipro^~/spack(xsdk-0.3.0) $ /Users/balay/spack/lib/spack/env/clang/clang -v conftest.c 
Apple LLVM version 9.0.0 (clang-900.0.38)
Target: x86_64-apple-darwin17.0.0
Thread model: posix
InstalledDir: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin
 "/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/clang" -cc1 -triple x86_64-apple-macosx10.13.0 -Wdeprecated-objc-isa-usage -Werror=deprecated-objc-isa-usage -emit-obj -mrelax-all -disable-free -disable-llvm-verifier -discard-value-names -main-file-name conftest.c -mrelocation-model pic -pic-level 2 -mthread-model posix -mdisable-fp-elim -fno-strict-return -masm-verbose -munwind-tables -target-cpu penryn -target-linker-version 302.3 -v -dwarf-column-info -debugger-tuning=lldb -resource-dir /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/lib/clang/9.0.0 -fdebug-compilation-dir /Users/balay/spack -ferror-limit 19 -fmessage-length 211 -stack-protector 1 -fblocks -fobjc-runtime=macosx-10.13.0 -fencode-extended-block-signature -fmax-type-align=16 -fdiagnostics-show-option -fcolor-diagnostics -o /var/folders/b2/4wzkzbgs3d91kqp1dw47wq940000gn/T/conftest-e0ba81.o -x c conftest.c
clang -cc1 version 9.0.0 (clang-900.0.38) default target x86_64-apple-darwin17.0.0
ignoring nonexistent directory "/usr/include"
#include "..." search starts here:
#include <...> search starts here:
 /usr/local/include
 /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/lib/clang/9.0.0/include
 /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include
 /System/Library/Frameworks (framework directory)
 /Library/Frameworks (framework directory)
End of search list.
conftest.c:1:10: fatal error: 'stdio.h' file not found
#include <stdio.h>
         ^~~~~~~~~
1 error generated.
balay@ipro^~/spack(xsdk-0.3.0) $ /usr/bin/clang -v conftest.c 
Apple LLVM version 9.0.0 (clang-900.0.38)
Target: x86_64-apple-darwin17.0.0
Thread model: posix
InstalledDir: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin
 "/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/clang" -cc1 -triple x86_64-apple-macosx10.13.0 -Wdeprecated-objc-isa-usage -Werror=deprecated-objc-isa-usage -emit-obj -mrelax-all -disable-free -disable-llvm-verifier -discard-value-names -main-file-name conftest.c -mrelocation-model pic -pic-level 2 -mthread-model posix -mdisable-fp-elim -fno-strict-return -masm-verbose -munwind-tables -target-cpu penryn -target-linker-version 302.3 -v -dwarf-column-info -debugger-tuning=lldb -resource-dir /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/lib/clang/9.0.0 -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.13.sdk -I/usr/local/include -fdebug-compilation-dir /Users/balay/spack -ferror-limit 19 -fmessage-length 211 -stack-protector 1 -fblocks -fobjc-runtime=macosx-10.13.0 -fencode-extended-block-signature -fmax-type-align=16 -fdiagnostics-show-option -fcolor-diagnostics -o /var/folders/b2/4wzkzbgs3d91kqp1dw47wq940000gn/T/conftest-7ccbaa.o -x c conftest.c
clang -cc1 version 9.0.0 (clang-900.0.38) default target x86_64-apple-darwin17.0.0
ignoring nonexistent directory "/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.13.sdk/usr/local/include"
ignoring nonexistent directory "/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.13.sdk/Library/Frameworks"
#include "..." search starts here:
#include <...> search starts here:
 /usr/local/include
 /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/lib/clang/9.0.0/include
 /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include
 /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.13.sdk/usr/include
 /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.13.sdk/System/Library/Frameworks (framework directory)
End of search list.
 "/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/ld" -demangle -lto_library /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/lib/libLTO.dylib -no_deduplicate -dynamic -arch x86_64 -macosx_version_min 10.13.0 -syslibroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.13.sdk -o a.out /var/folders/b2/4wzkzbgs3d91kqp1dw47wq940000gn/T/conftest-7ccbaa.o -L/usr/local/lib -lSystem /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/lib/clang/9.0.0/lib/darwin/libclang_rt.osx.a
balay@ipro^~/spack(xsdk-0.3.0) $ 

```

Looks like spack clang wrapper is setting up some env variable that is preventing /usr/bin/clang from using the following options. /usr/bin/clang is using them by default.

```
-isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.13.sdk -I/usr/local/include
```

Ah - spack is using /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/clang - and not /usr/bin/clang

These appear to behave differently.

```
balay@ipro^~/spack(xsdk-0.3.0) $ /usr/bin/clang conftest.c 
balay@ipro^~/spack(xsdk-0.3.0) $ /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/clang conftest.c 
conftest.c:1:10: fatal error: 'stdio.h' file not found
#include <stdio.h>
         ^~~~~~~~~
1 error generated.
```

How do I tell spack to use /usr/bin/clang?

I tried modifying ~/.spack/darwin/compilers.yaml - but spack is still picking up the wrong one.

```
balay@ipro^~/spack(xsdk-0.3.0) $ cat ~/.spack/darwin/compilers.yaml
compilers:
- compiler:
    environment: {}
    extra_rpaths: []
    flags: {}
    modules: []
    operating_system: highsierra
    paths:
      cc: /usr/bin/clang
      cxx: /usr/bin/clang++
      f77: /usr/local/bin/gfortran
      fc: /usr/local/bin/gfortran
    spec: clang@9.0.0-apple
    target: x86_64
- compiler:
    environment: {}
    extra_rpaths: []
    flags: {}
    modules: []
    operating_system: highsierra
    paths:
      cc: /usr/local/bin/gcc-7
      cxx: /usr/local/bin/g++-7
      f77: /usr/local/bin/gfortran-7
      fc: /usr/local/bin/gfortran-7
    spec: gcc@7.2.0
    target: x86_64
balay@ipro^~/spack(xsdk-0.3.0) $ sh -x /Users/balay/spack/lib/spack/env/clang/clang -c conftest.c  2>&1 | grep clang | grep conftest
+ exec /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/clang -c conftest.c
```

```
balay@ipro^~/spack(xsdk-0.3.0) $ echo $SPACK_CC

balay@ipro^~/spack(xsdk-0.3.0) $ spack env pkg-config bash
balay@ipro^~/spack(xsdk-0.3.0) $ echo $SPACK_CC
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/clang

```

```

--- a/lib/spack/spack/compiler.py
+++ b/lib/spack/spack/compiler.py
@@ -125,7 +125,6 @@ class Compiler(object):
         def check(exe):
             if exe is None:
                 return None
-            exe = self._find_full_path(exe)
             _verify_executables(exe)
             return exe
 
@@ -285,17 +284,6 @@ class Compiler(object):
         # does not spoil the intented precedence.
         successful.reverse()
         return dict(((v, p, s), path) for v, p, s, path in successful)
-
-    def _find_full_path(self, path):
-        """Return the actual path for a tool.
-
-        Some toolchains use forwarding executables (particularly Xcode-based
-        toolchains) which can be manipulated by external environment variables.
-        This method should be used to extract the actual path used for a tool
-        by finding out the end executable the forwarding executables end up
-        running.
-        """
-        return path
 
     def setup_custom_environment(self, pkg, env):
         """Set any environment variables necessary to use the compiler."""

--- a/lib/spack/spack/compilers/clang.py
+++ b/lib/spack/spack/compilers/clang.py
@@ -158,16 +158,6 @@ class Clang(Compiler):
 
         return _version_cache[comp]
 
-    def _find_full_path(self, path):
-        basename = os.path.basename(path)
-
-        if not self.is_apple or basename not in ('clang', 'clang++'):
-            return super(Clang, self)._find_full_path(path)
-
-        xcrun = Executable('xcrun')
-        full_path = xcrun('-f', basename, output=str)
-        return full_path.strip()
-
     @classmethod
     def fc_version(cls, fc):
         # We could map from gcc/gfortran version to clang version, but on macOS
@@ -201,6 +191,14 @@ class Clang(Compiler):
             # with Clang. Those point to Spack's compiler wrappers and
             # consequently render MPI non-functional outside of Spack.
             return
+
+        # Use special XCode versions of compiler wrappers when using XCode
+        # Overwrites build_environment's setting of SPACK_CC and SPACK_CXX
+        xcrun = Executable('xcrun')
+        xcode_clang = xcrun('-f', 'clang', output=str).strip()
+        xcode_clangpp = xcrun('-f', 'clang++', output=str).strip()
+        env.set('SPACK_CC', xcode_clang, force=True)
+        env.set('SPACK_CXX', xcode_clangpp, force=True)
 
         xcode_select = Executable('xcode-select')
 

--- a/lib/spack/spack/environment.py
+++ b/lib/spack/spack/environment.py
@@ -460,7 +460,9 @@ def set_or_unset_not_first(variable, changes, errstream):
     modifications have already been requested.
     """
     indexes = [ii for ii, item in enumerate(changes)
-               if ii != 0 and type(item) in [SetEnv, UnsetEnv]]
+               if ii != 0 and
+               not item.args.get('force', False) and
+               type(item) in [SetEnv, UnsetEnv]]
     if indexes:
         good = '\t    \t{context} at {filename}:{lineno}'
         nogood = '\t--->\t{context} at {filename}:{lineno}'

